Fobnail blog post - 2 phase (#453)
3mdeb/news-and-ideas: blog/content/post/2021-12-10-fobnail_2nd_phase.md

---
title: Fobnail Token - developing communication method that meets the CHARRA requirements

> abstract: 'The Fobnail Token is an open-source hardware USB device that help to

Reviewer: ```suggestion
abstract: 'The Fobnail Token is an open-source hardware USB device that helps to
```

> abstract: 'The Fobnail Token is an open-source hardware USB device that help to

Reviewer 2: helps

           last phase, we managed to implement the communication method that
           will be used between verifier and attester.'
cover: /covers/image-file.png
author: tomasz.zyjewski
layout: post
published: true
date: 2021-12-10
archives: "2021"

tags:
  - fobnail
  - firmware
  - usb
  - CHARRA
categories:
  - Firmware
  - Security

---

# About the Fobnail Token project

The Fobnail Token is a project that aims to provide a reference architecture for
building offline integrity measurement servers on the USB device and clients
running in Dynamically Launched Measured Environments (DLME). It allows the
Fobnail owner to verify the trustworthiness of the running system before
performing any sensitive operation. This project was founded by [NlNet
Foundation](https://nlnet.nl/). More information about the project can be found
in the [Fobnail documentation](https://fobnail.3mdeb.com/).

# Communication in CHARRA

[CHARRA](https://github.com/Fraunhofer-SIT/charra) is a "Challenge/Response
Remote Attestation" interaction model of the IETF RATS Reference Interaction
Models for Remote Attestation Procedures using TPM 2.0. In this project the

> attester and verifier communnicate with themselfes using libcoap. In order to

Reviewer: ```suggestion
attester and verifier communnicate with themselves using libcoap. In order to
```

> attester and verifier communnicate with themselfes using libcoap. In order to

Reviewer 2: themselves

achieve that we need to implement Ethernet over USB on the Fobnail Token. We
decided to use Rust so [nrf-hal](https://github.com/nrf-rs/nrf-hal) project

> provide us with a USB driver, and the conducted

Reviewer: ```suggestion
provides us with a USB driver, and the conducted
```

[research](https://fobnail.3mdeb.com/eth-over-usb-research/)
allowed us to determine that EEM will be the most appropriate protocol

> implementing Ethernet over USB. Additionaly we use

Reviewer 2: Additionally,

[smoltcp](https://github.com/smoltcp-rs/smoltcp) which is an interesting project
that provides an implementation of TCP/IP stack.

# Building applications for Fobnail

We started our work on running the `hello-world` example using Rust nrf-hal. It
turns out that the repository is missing an example for the nRF52840 which we

> use as a Fobnail prototype. The needed code can be found on

Reviewer: @TomaszAIR This part is unclear. The seconds sentence mentions that it turns out that there is no example code for the platform we use, thats ok. But the next sentence says something about code, but it is unclear what is the needed code. We should mention here that we have ported the `blinky-demo` example to run on nRF52840, and then `The code can found in`

[Fobnail's](https://github.com/fobnail/nrf-hal/tree/blinky-demo-nrf52840/examples/blinky-demo-nrf52840)
fork of nrf-hal project. The full process is described in the
[documentation](https://fobnail.3mdeb.com/flashing_samples/).

> The next step was to implement EEM protocol and use it with smoltcp. The code

Reviewer: ```suggestion
The next step was to implement EEM protocol and integrate it with smoltcp. The code
```

can be found [here](https://github.com/fobnail/usbd-ethernet/tree/main/src).

> Like in the `hello-world` example, also here we use [dockerized Fobnail

Reviewer: Not sure about that, but the version below sounds better for me

```suggestion
Like in the `hello-world` example, here also we use [dockerized Fobnail
```

SDK](https://github.com/fobnail/fobnail-sdk) which allows to build Rust
applications. During the development, we encountered some
[problems](https://fobnail.3mdeb.com/implementing-eth-over-usb/#encountered-problems)
and the [status of the current
implementation](https://fobnail.3mdeb.com/implementing-eth-over-usb/#status-of-current-implementation)
can be found in Fobnail documentation.

The last step was to prepare a Fobnail firmware example, which for now is an
application that allows to read Ethernet frames and send them back unchanged
using the USB over Ethernet driver. Code is available
[here](https://github.com/fobnail/fobnail/blob/main/src/main.rs).

# Running Fobnail firmware

Running the Fobnail demo on the [nRF52840
dongle](https://www.nordicsemi.com/Products/Development-hardware/nrf52840-dongle)
is really straightforward if only the
[environment](https://fobnail.3mdeb.com/environment/) was correctly prepared.
[Tests](https://fobnail.3mdeb.com/implementing-eth-over-usb/#testing) results
have been made publicly available.

The Fobnail firmware was also prepared to [develop on
PC](https://fobnail.3mdeb.com/local_development/) thanks to that any hardware

> except a development machine is not needed for development.

Reviewer: @TomaszAIR I would change that to
```
The Fobnail firmware can also run directly on PC (see `Developing firmware on PC`), thanks to that it is possible to develop firmware without any additional hardware.
```

## Summary

As part of the described phase, we were able to implement Ethernet over USB and
properly run it on [nRF52840
dongle](https://www.nordicsemi.com/Products/Development-hardware/nrf52840-dongle).
It is also worth paying attention to the provided code that allows you to use
this implementation in isolation from the hardware layer - without using the USB
standard. This will allow you to work on CHARRA functionality in the future

> without the need for hardware. In the future, this project will be developed,

Reviewer: ```suggestion
without the need for hardware. Future development of this project will be presented in subsequent blog posts.
```
